src/hotspot/cpu/aarch64/frame_aarch64.cpp

```diff
@@ -507,8 +507,11 @@ bool frame::is_interpreted_frame_valid(JavaThread* thread) const {
   // do some validation of frame elements
 
   // first the method
-
-  Method* m = *interpreter_frame_method_addr();
+  Method** m_addr = interpreter_frame_method_addr();
+  if (m_addr == nullptr || !os::is_readable_pointer(m_addr)) {
+    return false;
+  }
+  Method* m = *m_addr;
```

Comment:
```suggestion
  Method* m_addr = interpreter_frame_method_addr();
  if (m_addr == nullptr) {
    return false;
  }
  Method* m = SafeFetchN(m_addr, nullptr);
  if (m == nullptr) {
    return false;
  }
```

Reason: more robust against changes in memory map, faster.

Reply:
Thanks for raising this. It doesn't look like `SafeFetchN` is async signal safe because it calls into this:

```cpp
template <class T>
ATTRIBUTE_NO_ASAN static bool _SafeFetchXX_internal(const T *adr, T* result) {

  T n = 0;

  // Set up a jump buffer. Anchor its pointer in TLS. Then read from the unsafe address.
  // If that address was invalid, we fault, and in the signal handler we will jump back
  // to the jump point.
  sigjmp_buf jb;
  if (sigsetjmp(jb, 1) != 0) {
    // We faulted. Reset TLS slot, then return.
    pthread_setspecific(g_jmpbuf_key, nullptr);
    *result = 0;
    return false;
  }

  // Anchor jump buffer in TLS
  pthread_setspecific(g_jmpbuf_key, &jb);

  // unsafe access
  n = *adr;

  // Still here... All went well, adr was valid.
  // Reset TLS slot, then return result.
  pthread_setspecific(g_jmpbuf_key, nullptr);
  *result = n;

  return true;

}
```

This means we shouldn't be calling `os:: is_readable_pointer` either here, because it calls `SafeFetch32` which calls into the same function above. I had originally intended to simply perform a null check here (because that's the condition we've actually observed) and will push a change to revert to the null check.

Reply:
> Thanks for raising this. It doesn't look like `SafeFetchN` is async signal safe because it calls into this:
> 
> ```c++
> template <class T>
> ATTRIBUTE_NO_ASAN static bool _SafeFetchXX_internal(const T *adr, T* result) {
> 
> ```
> 
> This means we shouldn't be calling `os:: is_readable_pointer` either here, because it calls `SafeFetch32` which calls into the same function above. I had originally intended to simply perform a null check here (because that's the condition we've actually observed) and will push a change to revert to the null check.

That's not how `SafeFetchN` is normally implemented. On Linux and BSD it's more like

```
    # Support for intptr_t SafeFetchN(intptr_t* address, intptr_t defaultval);
    #
    #  x1 : address
    #  x0 : defaultval
SafeFetchN_impl:
_SafeFetchN_fault:
    ldr      x0, [x0]
    ret
_SafeFetchN_continuation:
    mov      x0, x1
    ret
```

which should be fine.

Reply:
@tstuefe implemented explicitly for being signal safe.

Reply:
OK let's go with your suggestion, thanks for explaining. I'm actually skeptical this can actually be a non-null bad pointer, as we've only seen this crash happen once, and the pointer was null in that instance. But this solution looks robust, so thanks for suggesting it.

Reply:
So you can safely fetch something that might be a method, and even if it really is a method and that can be verified, it might be a concurrently unloading method that is about to get clobbered or uncommiyted. Dereferencing and using such a Method is also subject to occasional crashing. Might be a problem for another day though.

Reply:
@richardstartin About async-safety: all supported (jit) architectures use static assembly as @theRealAph pointed out, these should be signal safe. The code snippet you found is only used by zero. You are probably not concerned with zero. And even there, yes, we longjmp out of signal handling, since there is no other way to implement SafeFetch in zero. That is technically async-sig-unsafe, but in practice it works and is tested for use in signal handlers.

About safety, @fisk is right in that this is still not completely safe since Method (and any of the objects chained to it that AGCT implicitly relies on being there, e.g. ConstMethod) can get out of scope while AGCT uses them.

Reply:
@fisk this might be solved in a new version of the API. I already spoke with @apangin about this and he's positive :)

Reply:
@tstuefe @fisk I hadn't appreciated that the cause was probably concurrent method unloading, we don't have a core dump, just the backtrace from the crash and the disassembly from objdump, so all I knew was that the pointer was null but not why. This is not the sort of thing that reproduces readily. I don't have as much context about the adjacent JVM mechanisms as others in this thread and am just trying to fix a crash based on the evidence I have.

This pointer being null seems to be a symptom rather than a cause and it doesn't appear there's anything we can do about concurrent method unloading interfering with AsyncGetCallTrace, so I wonder how worthwhile attempting to fix this is. On the one hand it will crash another way sometimes, on the other hand the probability of this happening is significantly reduced to the subsequent usages of the pointer, whereas that window of time for unloading a method to cause a crash in AsyncGetCallTrace is currently the duration of the unwind preceding the current frame. Let me know what you think about proceeding and I'll submit a fix with the null check which would have been sufficient to avoid the observed segfault. Thanks.

Reply:
@richardstartin I definitly think a patch in the "aph" form - first check for NULL, then check again with SafeFetch - makes a lot of sense. Not perfect, but will reduce the chance of crashes happening. And its very simple and backportable.

Reply:
Hardening the API is always a good idea, especially if it doesn't have a performance impact. We generally don't know in which state ASGCT is called. I added comparable checks at many places before.

Reply:
I prepared a new patch incorporating the suggestions made in this thread, which seems more straightforward to review given the force push and having gone round in circles #15193